From 9f77b6b25c5733c8431fc84b8a5df0f6179c26ec Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Wed, 26 Sep 2018 16:00:25 +0800
Subject: [PATCH 4742/5242] MLK-19737 drm/imx: dpu: crtc: Explicitly reject
 using PC for LVDS/DSI encoder

commit  6c8a2db26ae066783562df2a37f57da30369b411 from
https://source.codeaurora.org/external/imx/linux-imx.git

Users may specify any video mode for fbdev emulation in boot up command
line.  Pixel combiner is likely needed for that video mode.  However,
pixel combiner cannot work together with LVDS or DSI encoders for any
existing SoCs.  Thus, let's explicitly reject this case in ->atomic_check.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-crtc.c |   27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-crtc.c b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
index 556e21f..cbdd70f 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-crtc.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
@@ -404,6 +404,8 @@ static irqreturn_t dpu_dec_shdld_irq_handler(int irq, void *dev_id)
 static int dpu_crtc_atomic_check(struct drm_crtc *crtc,
 				 struct drm_crtc_state *crtc_state)
 {
+	struct drm_device *dev = crtc->dev;
+	struct drm_encoder *encoder;
 	struct drm_plane *plane;
 	struct drm_plane_state *plane_state;
 	struct dpu_plane_state *dpstate;
@@ -411,14 +413,29 @@ static int dpu_crtc_atomic_check(struct drm_crtc *crtc,
 	struct dpu_crtc_state *dcstate = to_dpu_crtc_state(imx_crtc_state);
 	struct drm_display_mode *mode = &crtc_state->adjusted_mode;
 	struct videomode vm;
+	unsigned long encoder_types = 0;
+	u32 encoder_mask;
 	int i = 0;
 
-	if (crtc_state->enable) {
-		drm_display_mode_to_videomode(mode, &vm);
+	list_for_each_entry(encoder, &dev->mode_config.encoder_list, head) {
+		encoder_mask = 1 << drm_encoder_index(encoder);
+
+		if (!(crtc_state->encoder_mask & encoder_mask))
+			continue;
 
-		if (dcstate->use_pc &&
-		    ((vm.hactive % 2)   || (vm.hfront_porch % 2) ||
-		     (vm.hsync_len % 2) || (vm.hback_porch % 2)))
+		encoder_types |= BIT(encoder->encoder_type);
+	}
+
+	if (crtc_state->enable && dcstate->use_pc) {
+		if (encoder_types & BIT(DRM_MODE_ENCODER_LVDS))
+			return -EINVAL;
+
+		if (encoder_types & BIT(DRM_MODE_ENCODER_DSI))
+			return -EINVAL;
+
+		drm_display_mode_to_videomode(mode, &vm);
+		if ((vm.hactive % 2)   || (vm.hfront_porch % 2) ||
+		    (vm.hsync_len % 2) || (vm.hback_porch % 2))
 			return -EINVAL;
 	}
 
-- 
1.7.9.5

